### PR TITLE
TRPO Implementation

### PR DESCRIPTION
**Pull request by plutasnyy**
## What does this PR do?
Hello! As discussed in [issue 596](https://github.com/PyTorchLightning/lightning-bolts/issues/596#issuecomment-922752890) together with @Naiwo we implemented the [TRPO](https://arxiv.org/abs/1502.05477) method in the RL module.

### Implementation & Results

In order to validate the implementation and prove correctness, our code was quite heavily based on an implementation from the [Implementation Matters in Deep Policy Gradients: A Case Study on PPO and TRPO](https://arxiv.org/abs/2005.12729) publication by Logan Engstrom et al. with source code in [Github](https://github.com/MadryLab/implementation-matters). 

We tested our code on environments from the publication (Walker, Hopper and Humanoid from MuJoCo, all are continuous), 
and one simple discrete, just to verify the correctness of the implementation - LunarLander.

Of course, our implementations differ in several ways - we set the direction of improvement based on the whole batch 
rather than 10%, different actor and critic architectures, training in a multi/single agent framework etc. 
The point here is not to reproduce the results exactly, but simply to show that our implementation meets the expected minimum ;)

The scores presented by the authors are aggregated as expected returns based on agent performance after 500 learning epochs. 
We simply show the average of the last 100 episodes after epochs execution (typically one epoch has significantly fewer 
episodes 2-10 depending on the environment).

##### Humanoid
Expected: [576, 596]
Ours - without problem to obtain cumulated rewards around 600

Command to reproduce: 
```bash
python3 trpo_model.py --max-episode-len 200 --lr-critic 1e-4 --kl-div-threshold 0.07 --env Humanoid-v2  --max_epochs 500 --log_every_n_steps 10
```

##### Hopper
Expected: [1948, 2136]
Ours - on the average the model works similarly

Command to reproduce: 
```bash
python3 trpo_model.py --max-episode-len 2048 --lr-critic 1e-3 --kl-div-threshold 0.05 --env Hopper-v2  --max_epochs 500 --log_every_n_steps 10
```

##### Walker
Expected: [2709, 2873]
Ours - Our model performs worse, even with more experience, however, with tuned parameters it meets the expected threshold

Command to reproduce: 
```bash
python3 trpo_model.py --max-episode-len 2048 --lr-critic 1e-3 --kl-div-threshold 0.1 --env Walker2d-v2  --max_epochs 1000 --log_every_n_steps 10
```

##### LunarLander
Expected - in short, a score around 200 indicates a correctly completed task
Ours - the agent was able to learn how to solve task

Command to reproduce: 
```bash
python3 trpo_model.py --max-episode-len 2048 --lr-critic 1e-3 --kl-div-threshold 0.1 --env LunarLander-v2  --max_epochs 1000 --log_every_n_steps 10
```

You can see more plots in the link: [wandb](https://wandb.ai/ghostteam/trpo/reports/Training-TRPO--VmlldzoxNDUyOTQ4).

If you would like to run our code without installing MuJoCo you can use [Colab](https://colab.research.google.com/drive/1LJMd5zT9hjtuJGf1SBXzwzdBp8fNgrb1?usp=sharing), which we also used for a while ;)

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does **only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? - Yes, but please check if we haven't missed something
- [x] Did you write any **new necessary tests**? \[not needed for typos/docs\]
- [x] Did you verify new and **existing tests** pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-bolts/blob/master/CHANGELOG.md)?

## PR review

- [x] Is this pull request **ready for review**? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Absolutely!

### CODE REVIEW

**Comment by plutasnyy** [comment]
We'll fix https://deepsource.io/gh/PyTorchLightning/lightning-bolts/run/fbfd5cea-cbfd-4e27-9dc9-ffd28658dcb1/python/ in a free moment

**Comment by Borda** [comment]
> I noticed that unfortunately new models are no longer accepted

RL is still welcome, we are just limiting SSL contribution as we are rethinking/reorganizing this part...
also, please bare with us as recently we are a bit short with time... in contrast, if you are interested to take a larger contribution stake, pls ping me on slack, and happy to chat about options :)
